src/bun.js/api/bun/udp_socket.zig

```diff
@@ -0,0 +1,751 @@
+        const addr_ptr = brk: {
+            if (dst) |dest| {
+                if (!this.parseAddr(globalThis, dest.port, dest.address, &addr)) {
+                    globalThis.throwInvalidArguments("Invalid address", .{});
+                    return .zero;
+                }
+                break :brk &addr;
+            } else {
+                break :brk null;
+            }
+        };
+
+        const res = this.socket.send(&.{payload.ptr}, &.{payload.len}, &.{addr_ptr});
+        if (bun.JSC.Maybe(void).errnoSys(res, .send)) |err| {
+            globalThis.throwValue(err.toJS(globalThis));
+            return .zero;
+        }
+        return JSValue.jsBoolean(res > 0);
+    }
+
+    fn parseAddr(
+        this: *This,
+        globalThis: *JSGlobalObject,
+        port_val: JSValue,
+        address_val: JSValue,
+        storage: *std.os.sockaddr.storage,
+    ) bool {
+        _ = this;
+        const number = port_val.coerceToInt32(globalThis);
+        const port: u16 = if (number < 1 or number > 0xffff) 0 else @intCast(number);
+
+        const str = address_val.toBunString(globalThis);
+        defer str.deref();
+        const address_slice = str.toOwnedSliceZ(default_allocator) catch bun.outOfMemory();
+        defer default_allocator.free(address_slice);
+
+        var addr4: *std.os.sockaddr.in = @ptrCast(storage);
+        if (inet_pton(std.os.AF.INET, address_slice.ptr, &addr4.addr) == 1) {
+            addr4.port = htons(@truncate(port));
+            addr4.family = std.os.AF.INET;
+        } else {
+            var addr6: *std.os.sockaddr.in6 = @ptrCast(storage);
+            if (inet_pton(std.os.AF.INET6, address_slice.ptr, &addr6.addr) == 1) {
+                addr6.port = htons(@truncate(port));
+                addr6.family = std.os.AF.INET6;
+            } else {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    const Destination = struct {
+        port: JSValue,
+        address: JSValue,
+    };
+
+    pub fn ref(this: *This, globalThis: *JSC.JSGlobalObject, _: *JSC.CallFrame) callconv(.C) JSValue {
+        if (!this.closed) {
+            this.poll_ref.ref(globalThis.bunVM());
+        }
+
+        return .undefined;
+    }
+
+    pub fn unref(this: *This, globalThis: *JSC.JSGlobalObject, _: *JSC.CallFrame) callconv(.C) JSValue {
+        if (!this.closed) {
+            this.poll_ref.unref(globalThis.bunVM());
+        }
+
+        return .undefined;
+    }
+
+    pub fn close(
+        this: *This,
+        globalThis: *JSGlobalObject,
+        _: *CallFrame,
+    ) callconv(.C) JSValue {
+        if (this.closed) {
+            globalThis.throw("Socket is already closed", .{});
+            return .zero;
+        }
+
+        this.doClose();
+
+        return .undefined;
+    }
+
+    fn doClose(
+        this: *This,
+    ) void {
+        this.closed = true;
+        this.config.unprotect();
```

Comment:
Can the socket be closed from usockets due to an error? Can the `onClose` callback still be called? 

I don't think it's safe to clean up the socket handlers until after usockets calls the on close callback